src/fontra/views/fontinfo/panel-sources.js

```diff
@@ -1,14 +1,604 @@
+    // but makes more sense directly in the glyph editing window.
+  }
+
+  prepareVerticalMetricsForController(verticalMetrics) {
+    let newVerticalMetrics = {};
+    for (const key in verticalMetrics) {
+      newVerticalMetrics[`value-${key}`] = verticalMetrics[key].value;
+      newVerticalMetrics[`zone-${key}`] = verticalMetrics[key].zone;
+    }
+    return newVerticalMetrics;
+  }
+
+  checkSourceLocation(axisName, value) {
+    const newLocation = { ...this.source.location, [axisName]: value };
+    return this.checkSourceEntry("location", undefined, newLocation);
+  }
+
+  checkSourceEntry(key, valueKey = undefined, value) {
+    let errorMessage = "";
+    for (const sourceIdentifier in this.sources) {
+      if (sourceIdentifier == this.sourceIdentifier) {
+        // skip the current source
+        continue;
+      }
+      const source = this.sources[sourceIdentifier];
+
+      let existsAlready = false;
+      let sourceValue;
+
+      if (valueKey == undefined) {
+        if (key == "location") {
+          sourceValue = locationToString(source[key]);
+          value = locationToString(value);
+        } else {
+          sourceValue = source[key];
+        }
+      } else {
+        sourceValue = source[key][valueKey];
+      }
+
+      if (sourceValue == value) {
+        existsAlready = true;
+      }
+
+      if (existsAlready) {
+        errorMessage = `${key}${valueKey ? " " + valueKey : ""}: “${value}”
+          exists already, please use a different value.`;
+        break;
+      }
+    }
+
+    if (errorMessage) {
+      message(`Can’t edit font source`, errorMessage);
+      this.setupUI();
+      return false;
+    }
+    return true;
+  }
+
+  editSource(editFunc, undoLabel) {
+    const root = { sources: this.sources };
+    const changes = recordChanges(root, (root) => {
+      editFunc(root.sources[this.sourceIdentifier]);
+    });
+    if (changes.hasChange) {
+      this.postChange(changes.change, changes.rollbackChange, undoLabel);
+    }
+  }
+
+  deleteSource() {
+    const undoLabel = `delete source '${this.source.name}'`;
+    const root = { sources: this.sources };
+    const changes = recordChanges(root, (root) => {
+      delete root.sources[this.sourceIdentifier];
+    });
+    if (changes.hasChange) {
+      this.postChange(changes.change, changes.rollbackChange, undoLabel);
+      this.setupUI();
+    }
+  }
+
+  toggleShowHide() {
+    const el = this.querySelector("#open-close-icon");
+    el.style.transform =
+      el.style.transform === "rotate(180deg)" ? "rotate(0deg)" : "rotate(180deg)";
+
+    this.style.maxHeight = this.style.maxHeight === "100%" ? "70px" : "100%";
+    for (const child of this.children) {
+      if (!child.style.maxHeight) {
+        child.style.maxHeight = "100%";
+      } else {
+        delete child.style;
+      }
+    }
+  }
+
+  _updateContents() {
+    const models = this.models;
+
+    for (const key in models) {
```

Comment:
Currently, `models` isn't _really_ a customizable / infinitely variable set of models. We literally have only three keys:
- general
- location
- verticalMetrics

In the code below, you add a generic listener callback to _all_ items, and the _inside_ the listener you switch again by key. This is inside out, and we're better off assigning the listeners explicitly. The code will be easier to understand, and possibly shorter as there's no need for all those if statements.

So, indeed use a loop to create the controllers (they are generic for each model), but then add listeners per item explicitly.

For example:
```js
    this.controllers.general.addListener((event) => {
      // listener code for "general"
    }
    this.controllers.location.addListener((event) => {
      // listener code for "location"
    }
    // etc.
```

Reply:
In the case of `if (event.key == "name") {` you can use `addKeyListener("name", (event) => ...)` instead, and save another if.